### DISCUSSION TITLE
rpc: Add test-only RPCs under `-test=<option>` flag

### DISCUSSION THREAD

**Discussion by Prabhat1308** [Bitshala/BitcoinCore-PR-Review-Club#83] [PR Reviews]
## Session Details
Date:05-09-2024
Time: IST 20:00 (UTC 14:30)
Link : https://github.com/bitcoin/bitcoin/pull/30717
Difficulty : Medium

## Notes
This PR splits off test-only RPCs to be included in -test flag with additional restrictions of use in regtest segregating the user facing and testing RPCs . These test-only RPCs are split off from args and have their own method to extract arguments .

## Questions
1. What is the need to differentiate normal RPCs and test RPCs ?
2. How do we identify dev-only RPC ?
3. What is the need to limit the `-test` RPC to `regtest` and not other testnets ?

## Learning
RPC 
tests

**Comment by paperpsych** [comment]
@VirdiJatin ... And @Prabhat1308 please confirm if the tags are good to go!

![Bitshala RC (30717)](https://github.com/user-attachments/assets/8d5c1d92-e298-4eac-bbcd-c644ca7d0145)

**Comment by Prabhat1308** [comment]
@paperpsych these are good to go.

**Comment by Prabhat1308** [comment]
### What is the need to differentiate normal flags and test flags ?
Test flags are experimental and needs to be used with the utmost care as it could lead to breaking the software . Normal flags are tailored for use by general audience and are battle tested. Currently bitcoin has no restriction put on these test flags and these can be used accidently by a normal user which can be a hazardous situation.
### How do we identify dev-only glags ?
The rpcs that are listed with `DEBUG_ONLY` and `DEBUG_TEST` are generally termed as dev-only flags . There is currently no restriction on these rpcs to only being used by devs in the bitcoin-core repo.
### What is the need to limit the -test flag to regtest and not other testnets ?
Since regtest is only user-specific and don't need interactions from an external party , It is considered safer and hence the need for limiting it to regtest. However , testing is required to be done in an environment that is closer to the mainnet and other testnets are better in that so this decision could be debated.
